### PR TITLE
Documentation Update #21

### PR DESCRIPTION
**Pull request by udaysinghcode**
First draft of documentation - I've written the intro and going through the theory of each method. I'm going through each method and will be writing more on how they all work to fill up the entire breadth of the design document. #21

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,13 +1,408 @@
...
 
+Languages are used as tools to communicate, and incidentally, the structure and limitations of a C++ determine the way we think. Learning how to think and within the context of a language provides one the ability to be a proper user, well at the same time 
+
+> In 24 hours you might be able to learn some of the syntax of C++ (if you already know another language), but you couldn't learn much about how to use the language. In short, if you were, say, a Basic programmer, you could learn to write programs in the style of Basic using C++ syntax, but you couldn't learn what C++ is actually good (and bad) for. So what's the point? Alan Perlis once said: "A language that doesn't affect the way you think about programming, is not worth knowing" -- [Peter Norvig][1]
+
+[1]:http://www.norvig.com/21-days.html
+
+This is the principle of linguistic relativity also known as the [Sapir-Whorf hypothesis][2] shows that the constraints of a language are what most often limit our thought processing. This idea, when applied to programming languages shows how Lisp programmers use macros for everything, Python programmers even in Ruby have code that looks like python.
```
Comment (natebrennand) [inline README.md]: the last sentence is confusing here

File: README.md [hunk 2]
```diff
@@ -1,13 +1,408 @@
...
+
+[2]:http://en.wikipedia.org/wiki/Linguistic_relativity
+
+One of the key components to developing C++ is learnability or more importantly, how easy is it for users to accomplish basic tasks. 
+
+C++ Collections is a C++ 11 library that provides lazy evaluation and functional transformations of the data to lighten the burden of using functional operations such as map, range, and filter for the user. To use C++ Streams, download them from Github and simply *#include collections.h* in the header of your file after dragging it in.
+
+----
+## What is C++ Collections?
+C++ Collections provides abstractions on a set of ordered data. All examples below assume that we are `using namespace cpp_collections`.
```
Comment (natebrennand) [inline README.md]: I think the second sentence should be moved to a footnote or something along those lines.
Just leave a comment to make it a footnote and we can do that when we transition to latex.

File: README.md [hunk 3]
```diff
@@ -1,13 +1,408 @@
...
+```
+
+The syntax you will note is similar to both Haskell and Scala.
+
+```haskell
+sum = foldl1 (+) [1..100]
+```
+
+```scala
+val sum = Range(1, 100).reduce((a,b) => a+b)
```
Comment (natebrennand) [inline README.md]: s/100/101

File: README.md [hunk 4]
```diff
@@ -1,13 +1,408 @@
...
+```
+
+The main abstraction introduced by this library is Collection. The capitalization om the word 'Collection' is important to distinguish it from collection. The Collection is different from standard collections in C++ in that:
```
Comment (natebrennand) [inline README.md]: s/om/in

File: README.md [hunk 5]
```diff
@@ -1,13 +1,408 @@
...
+
+1. It is functional in nature. All operations on a Collection produce some form of result by modifying the source in memory.
```
Comment (natebrennand) [inline README.md]: Functional means that it isn't modifying the source in place. It's returning a modified copy.
The point is that once you declare X, you can't modify X without reassigning it.

File: README.md [hunk 6]
```diff
@@ -1,13 +1,408 @@
...
+```
+
+```scala
+val sum = Range(1, 100).reduce((a,b) => a+b)
+```
+
+The main abstraction introduced by this library is Collection. The capitalization om the word 'Collection' is important to distinguish it from collection. The Collection is different from standard collections in C++ in that:
+
+1. It is functional in nature. All operations on a Collection produce some form of result by modifying the source in memory. 
+2. It is concise. After introducing lambda functions in C++11, and functors long before that, C++ has emerged as a viable language in which to write programs that conform to the functional paradigm. There is no template overhead to calling functions.
```
Comment (natebrennand) [inline README.md]: What do you mean by "no template overhead"?

File: README.md [hunk 7]
```diff
@@ -1,13 +1,408 @@
...
+3. Laziness seeking. We use laziness to optimize many Collections operations. So calling slice on the first five consecutive numbers will not require all numbers in the list to be examined.
```
Comment (natebrennand) [inline README.md]: Should this be "Lazy Evaluation"? I think you should acknowledge that this only applies to streams

File: README.md [hunk 8]
```diff
@@ -1,13 +1,408 @@
...
+```scala
+val sum = Range(1, 100).reduce((a,b) => a+b)
+```
+
+The main abstraction introduced by this library is Collection. The capitalization om the word 'Collection' is important to distinguish it from collection. The Collection is different from standard collections in C++ in that:
+
+1. It is functional in nature. All operations on a Collection produce some form of result by modifying the source in memory. 
+2. It is concise. After introducing lambda functions in C++11, and functors long before that, C++ has emerged as a viable language in which to write programs that conform to the functional paradigm. There is no template overhead to calling functions.
+3. Laziness seeking. We use laziness to optimize many Collections operations. So calling slice on the first five consecutive numbers will not require all numbers in the list to be examined. 
+4. Potentially unbounded. While all collections have a finite size, Collections do not. Short circuiting and infinite Collection by calling <section>head</section> on an infinite Collection.
```
Comment (natebrennand) [inline README.md]: Again, Collections have bounds. Streams do not

File: README.md [hunk 9]
```diff
@@ -1,13 +1,408 @@
...
+val sum = Range(1, 100).reduce((a,b) => a+b)
+```
+
+The main abstraction introduced by this library is Collection. The capitalization om the word 'Collection' is important to distinguish it from collection. The Collection is different from standard collections in C++ in that:
+
+1. It is functional in nature. All operations on a Collection produce some form of result by modifying the source in memory. 
+2. It is concise. After introducing lambda functions in C++11, and functors long before that, C++ has emerged as a viable language in which to write programs that conform to the functional paradigm. There is no template overhead to calling functions.
+3. Laziness seeking. We use laziness to optimize many Collections operations. So calling slice on the first five consecutive numbers will not require all numbers in the list to be examined. 
+4. Potentially unbounded. While all collections have a finite size, Collections do not. Short circuiting and infinite Collection by calling <section>head</section> on an infinite Collection. 
+[comment]: <> (Hopefully we can get infinite lists to work well)
```
Comment (natebrennand) [inline README.md]: you can use HTML style comments in markdown FYI

File: README.md [hunk 10]
```diff
@@ -1,13 +1,408 @@
...
+
+### Fluent Interfaces
+[Martin Fowler][mf] coined the term **fluent inteface** as a semantic facade that allows you to apply multiple properties to an object without having to redefine the object each time. Typically they are used to build domain specific languages, fluent interfaces provide not only more discoverable and readable code - but an easy structure to use particularly for people who may not be familiar with working with functional ideas in C++. Developing C++ Collections will benefit teams that would like to use method chaining as a replacement for typical collection based operations with syntax and code that is easier to parse and uses method chaining to communicate the actions taking place sequentially - similar to a UNIX pipe.
+
+[mf]:http://martinfowler.com/bliki/FluentInterface.html
```
Comment (natebrennand) [inline README.md]: can we consolidate the links to the top or bottom?

File: README.md [hunk 11]
```diff
@@ -1,13 +1,408 @@
...
+4. Potentially unbounded. While all collections have a finite size, Collections do not. Short circuiting and infinite Collection by calling <section>head</section> on an infinite Collection. 
+[comment]: <> (Hopefully we can get infinite lists to work well)
+
+### Fluent Interfaces
+[Martin Fowler][mf] coined the term **fluent inteface** as a semantic facade that allows you to apply multiple properties to an object without having to redefine the object each time. Typically they are used to build domain specific languages, fluent interfaces provide not only more discoverable and readable code - but an easy structure to use particularly for people who may not be familiar with working with functional ideas in C++. Developing C++ Collections will benefit teams that would like to use method chaining as a replacement for typical collection based operations with syntax and code that is easier to parse and uses method chaining to communicate the actions taking place sequentially - similar to a UNIX pipe.
```
Comment (natebrennand) [inline README.md]: s/inteface/interface

File: README.md [hunk 12]
```diff
@@ -1,13 +1,408 @@
...
+
+[mf]:http://martinfowler.com/bliki/FluentInterface.html
+
+This form of method cascading which originated in [Smalltalk][sm] allows a layer of syntactic sugar to be added to an otherwise repetitive piece of code. Cascasing prevents the need to list the object repeatedly as you can have numerous methods communicating exactly what they want to convey all acting on a single object by returning and propagating `this`.
```
Comment (natebrennand) [inline README.md]: s/Cascasing/Cascading

File: README.md [hunk 13]
```diff
@@ -1,13 +1,408 @@
...
+
+The main abstraction introduced by this library is Collection. The capitalization om the word 'Collection' is important to distinguish it from collection. The Collection is different from standard collections in C++ in that:
+
+1. It is functional in nature. All operations on a Collection produce some form of result by modifying the source in memory. 
+2. It is concise. After introducing lambda functions in C++11, and functors long before that, C++ has emerged as a viable language in which to write programs that conform to the functional paradigm. There is no template overhead to calling functions.
+3. Laziness seeking. We use laziness to optimize many Collections operations. So calling slice on the first five consecutive numbers will not require all numbers in the list to be examined. 
+4. Potentially unbounded. While all collections have a finite size, Collections do not. Short circuiting and infinite Collection by calling <section>head</section> on an infinite Collection. 
+[comment]: <> (Hopefully we can get infinite lists to work well)
+
+### Fluent Interfaces
```
Comment (natebrennand) [inline README.md]: i think this section could benefit from an example.

File: README.md [hunk 14]
```diff
@@ -1,13 +1,408 @@
...
+[mf]:http://martinfowler.com/bliki/FluentInterface.html
+
+This form of method cascading which originated in [Smalltalk][sm] allows a layer of syntactic sugar to be added to an otherwise repetitive piece of code. Cascasing prevents the need to list the object repeatedly as you can have numerous methods communicating exactly what they want to convey all acting on a single object by returning and propagating `this`.
+
+[sm]: http://en.wikipedia.org/wiki/Smalltalk
+
+
+## Design
+
+The C++ Collection class uses `std::vector<T>` to store data internally. This allows a number of different constructors to transfer information to the vector to control pointer management and data organization. Additionally, this allows lists, arrays, and C-styled arrays to be stored and mapped to the C++ Collection.
```
Comment (natebrennand) [inline README.md]: This is an implementation detail, not design.

File: README.md [hunk 15]
```diff
@@ -1,13 +1,408 @@
...
+This form of method cascading which originated in [Smalltalk][sm] allows a layer of syntactic sugar to be added to an otherwise repetitive piece of code. Cascasing prevents the need to list the object repeatedly as you can have numerous methods communicating exactly what they want to convey all acting on a single object by returning and propagating `this`.
+
+[sm]: http://en.wikipedia.org/wiki/Smalltalk
+
+
+## Design
+
+The C++ Collection class uses `std::vector<T>` to store data internally. This allows a number of different constructors to transfer information to the vector to control pointer management and data organization. Additionally, this allows lists, arrays, and C-styled arrays to be stored and mapped to the C++ Collection. 
+
+Within the Collection, there are three main types of operations which are elaborated below: source operators, intermediate operators, and terminal operators. Source operators allow the creation of Collections. Intermediate operators work off Collections and return Collections. Finally, Terminal operators return a non-Collection object or no object in return.
```
Comment (natebrennand) [inline README.md]: Why call them "operators"? I thought of them as constructors since that's what they're doing.

File: README.md [hunk 16]
```diff
@@ -1,13 +1,408 @@
...
+
+- Intermediate operators: These are Collections that both input and output streams. intermediate operators range between O(1) and O(n) depending on whether they are stateless or not.
+
+- Terminal operators: These are operators that close a Collection and finally cause all of the operations in the pipeline to evaluate and return a non-Collection value. 
+
+All Collection pipelines are formed by a combination of a source operator, an either singular or multiple intermediate operators, and concluded with a terminal operator. These three operators work together to form a pipeline. Unlike pipelines in Java 8, our pipelines are executed after each individual function, and therefore we do not need to read the terminal operation before our pipeline is executed. 
+
+Additionally, unlike Java 8, despite syntax similarities, we do not use consumabale Collections. The data of a Collection is statefully stored inside a vector in the Collection called `std::vector<T> Data`. A Collection is an abstraction on an `std::vector<T> Data`, so many lazily evaluated function require the use of memoization by increasing memory costs to decrease speed.
```
Comment (natebrennand) [inline README.md]: s/consumabale/consumable

File: README.md [hunk 17]
```diff
@@ -1,13 +1,408 @@
...
+
+As mentioned above, the example where one would slice the first five consecutive integers from a list of 1000 numbers involves a new vector being instaniated internally to transfer the first five data elements of the Collection to a newly created list which gets converted into a Collection. The theory behinnd many of the functions utilized for the Collections library will be expanded upon in the documentation below.
+
+
+Below we will tour through the design, the style, and way different methods function and operate.
+
+### Generators
+
+Generators are important in
```
Comment (natebrennand) [inline README.md]: Looks like you meant to finish this sentence.

File: README.md [hunk 18]
```diff
@@ -1,13 +1,408 @@
...
+```
+
+#### Collection<T>()
+
+**Method**:
+```cpp
+auto emptyCollection = Collection<T>();
+```
+
+Construct an empty collection of size 'size'. This Colleciton will return nothing but can be used with *Collection<T>.size()* to return a size.
```
Comment (natebrennand) [inline README.md]: s/Colleciton/Collection

File: README.md [hunk 19]
```diff
@@ -1,13 +1,408 @@
...
+### Generators
+
+Generators are important in 
+
+#### Collection<T>()
+Construct an empty collection. This Colleciton will return nothing but can be used with <section>Collection<T>.size()</section> to return a size of 0.
+
+*Example*:
+```cpp
+std::cout<< Collection<int>().size() << std::endl;
```
Comment (natebrennand) [inline README.md]: can you put a space between `std::cout` and `<<` in all of the snippets?

File: README.md [hunk 20]
```diff
@@ -1,13 +1,408 @@
...
+The second object in the C++ Collections library that we would like to expand on is the Stream object. Streams are developed by having a head of anytype, a tail which is initially set to a nullptr and a generator function. The initial stream allows users to run `take()`, `head()`, `tail()`, `filter()`, and `map()` on it. Yet Streams also allow the development of infinite lists using the `from()` as well. As the Stream object is developed, the documentation will be updated correspondingly. 
+
+## Development Suport
+
+If you are interested in contributing, 
+```
+g++ -std=c++11 -Wall tests/<FILE>.cpp
+```
+
+## Tests
```
Comment (natebrennand) [inline README.md]: I think this should be in a section called "Development Notes" / Remarks. A user of this wouldn't be running the test suite.

**Comment by natebrennand** [comment]
It'd be great if each sentence had it's own line. That makes it a lot easier to see edits. Markdown (and latex) will combine sentences into paragraphs as long as there's not a blank line between them.

File: README.md (hunk 1 above)
Comment (udaysinghcode) [inline README.md]: Yeah I'll clear it up, I was just trying to convey that languages have limitations in expression.

File: README.md (hunk 2 above)
Comment (udaysinghcode) [inline README.md]: Okay sounds good.

File: README.md (hunk 3 above)
Comment (udaysinghcode) [inline README.md]: Scala - Range(1,101)?

File: README.md (hunk 6 above)
Comment (udaysinghcode) [inline README.md]: Referring to the first issue we posted on GitHub. Essentially the Collection class is type agnostic, so it works with everything due to how it's written. I'll specify more.

File: README.md (hunk 7 above)
Comment (udaysinghcode) [inline README.md]: Yeah I had that in there - like 3 and 4 before I saw the Streams class - I was figuring we were working on it so I had it in place, but now that it's under the Streams class I'll split them off and list them separately.

File: README.md (hunk 8 above)
Comment (udaysinghcode) [inline README.md]: See note above^

File: README.md (hunk 9 above)
Comment (udaysinghcode) [inline README.md]: Oh actually? didn't know - as you can see this was before we had the Streams working - I'll fix the documentation to reflect it

File: README.md (hunk 10 above)
Comment (udaysinghcode) [inline README.md]: Yeah I'll consolidate them all on the bottom.

File: README.md (hunk 11 above)
Comment (udaysinghcode) [inline README.md]: noted.

File: README.md (hunk 12 above)
Comment (udaysinghcode) [inline README.md]: noted.

File: README.md (hunk 13 above)
Comment (udaysinghcode) [inline README.md]: Okay, I'll write up a syntax example and explain how it works. Not a problem.

File: README.md (hunk 15 above)
Comment (udaysinghcode) [inline README.md]: I was trying to think of some sort of unique word to look at them as three different things that represent a pipe structure. Would you prefer constructors or generators @natebrennand ?

File: README.md (hunk 17 above)
Comment (udaysinghcode) [inline README.md]: Oh I didn't even notice that. Okay thanks.

File: README.md (hunk 19 above)
Comment (udaysinghcode) [inline README.md]: yeah no worries

File: README.md (hunk 20 above)
Comment (udaysinghcode) [inline README.md]: Okay I'll put that all under something separate - that or we could do a separate document? I think just a section will work though.

**Comment by udaysinghcode** [comment]
Okay perfect, I'll do that as well. I wasn't really sure the best way to format the sentence structure, but that makes a lot of sense. Thanks for all of this, sometimes when you're just writing you miss spelling stuff, etc. I'll fix this up - and then just grind through the tutorial. I was just thinking to show how a number of different methods work and can stack together.

File: README.md (hunk 3 above)
Comment (ianzapolsky) [inline README.md]: he's saying you should replace the occurrences of 100 with 101 here

**Comment by natebrennand** [comment]
merging this and going to make all suggestions in a PR

**Pull request merged**